### ISSUE TITLE
mgr_.Init(traineddata_path.c_str()):Error:Assert failed:

### ISSUE THREAD

**Issue by iuriigalaida** [tesseract-ocr/tesseract#1075]
I've got error 'mgr_.Init(traineddata_path.c_str()):Error:Assert failed:in file ../../../../lstm/lstmtrainer.h, line 110' when trying to run a command below. Can somebody help me with this?
All paths are correct.
version 4.0

c:\Temp\tesseract-master\tesseract-master\api>lstmtraining --model_output C:/Temp/engbest/model --continue_from C:/Temp/engbest/eng.lstm --traineddata C:/Temp/engbest/eng.traineddata --old_traineddata C:/Temp/tesseract-master/tesseract-master/tessdata/eng.traineddata --train_listfile C:/Temp/engbest/eng.training_files.txt --max_iterations 3600

**Comment by stephenyong2005** [comment]
Do you put the traineddate in the write directory? 
--traineddata C:/Temp/engbest/eng.traineddata
Please check that if there is eng.traineddata in C:/Temp/engbest/? If no, copy the eng.traineddata to C:/Temp/engbest/. 

Then, I think this will be resolved. I have met the same problem, and this is my resolution.

**Comment by iuriigalaida** [comment]
It helped. Thanks. But than I got 

Loaded file C:/Temp/engbest/eng.lstm, unpacking...
Warning: LSTMTrainer deserialized an LSTMRecognizer!
Code range changed from 105 to 105!!
Failed to continue from: C:/Temp/engbest/eng.lstm

stephenyong2005 Did you see error like this in your case?

**Comment by stephenyong2005** [comment]
Sorry, I do not got such a error like you mentioned, but i got another error. By the way, I am a new user of tesseract. Do you know how to generate a lstm file? Thanks.

2017年8月13日 16:09，iuriigalaida <notifications@github.com>写道：

It helped. Thanks. But than I got

Loaded file C:/Temp/engbest/eng.lstm, unpacking...
Warning: LSTMTrainer deserialized an LSTMRecognizer!
Code range changed from 105 to 105!!
Failed to continue from: C:/Temp/engbest/eng.lstm

stephenyong2005 Did you see error like this in your case?

**Comment by Shreeshrii** [comment]
Are you using the latest source from github for building tesseract? What is your version info, git log info?

**Comment by Shreeshrii** [comment]
Also see https://github.com/tesseract-ocr/tesseract/issues/1069
regarding 'Failed to continue from' error

**Comment by stephenyong2005** [comment]
I use v4.0.0. Compile from the source code instead of git. OS is ubuntu 17.04. I have finished the demo procedure in the wiki web page of training tesseract 4.0 by following the tutorial.

2017年8月14日 00:18，Shreeshrii <notifications@github.com>写道：

Are you using the latest source from github for building tesseract? What is your version info, git log info?

**Comment by Shreeshrii** [comment]
> Compile from the source code  instead of git.

from where?

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by stephenyong2005** [comment]
this is the tesseract wiki homepage:  https://github.com/tesseract-ocr/tesseract/wiki

2017年8月14日 00:56，Shreeshrii <notifications@github.com>写道：
> Compile from the source code instead of git.

from where?

ShreeDevi
____________________________________________________________
भजन - कीर्तन - आरती @ http://bhajans.ramparivar.com

**Comment by iuriigalaida** [comment]
After I get latest sourses v4 lstmtraining started work but I almost always got errors like 
Encoding of string failed!
Can't encode transcription:
Even if I directly set path to unicharset file.

**Comment by Shreeshrii** [comment]
That means you have new characters in your training text, so fine tuning
may not work.

**Comment by iuriigalaida** [comment]
yes. i have 44 new character and I wont to extend english traineddata. Attached my unicharset file. How can I aviod this encoding issues?

**Comment by Shreeshrii** [comment]
Unicharset was not attached - 

see https://github.com/tesseract-ocr/tesseract/wiki/TrainingTesseract-4.00#training-just-a-few-layers

Have you tried Latin.traineddata - it will have more characters than eng in it.

**Comment by iuriigalaida** [comment]
[eng.zip](https://github.com/tesseract-ocr/tesseract/files/1222246/eng.zip)

**Comment by Shreeshrii** [comment]
Thanks for the unicharset. I notice it has accented english letters.

1. Test with Latin.traineddata - which is for Latin script not Latin
language (lat).

2. Check that all your additional characters are in best/Latin.traineddata
- use combine_tessdata -u to unpack the traineddata and look at its
unicharset.

I find it easy to sort a copy of the unicharsets and then compare .

3. If Latin.traineddata has all characters you want, then you can do
finetune using Latin.traineddata as the continue_from rather than English.

**Comment by Shreeshrii** [comment]
Re: 'mgr_.Init(traineddata_path.c_str()):Error:Assert failed:in file ../../../../lstm/lstmtrainer.h, line 110' 

This comes when the path to the traineddata files is incorrect. Check file locations and correct the command for training.

**Comment by amitdo** [comment]
The command should use tprintf() with a meaningful message and exit(), not assert.

**Comment by Shreeshrii** [comment]
@stweil for https://github.com/tesseract-ocr/tesseract/issues/1423

**Comment by amitdo** [comment]
https://github.com/tesseract-ocr/tesseract/blob/66b77e66390d5e641c3f84436a27bb469cad15a6/src/training/unicharset/lstmtrainer.h#L99

**Comment by zdenop** [comment]
What is expected solution here? Just note that path does not exist?
 
```diff
diff --git a/src/training/unicharset/lstmtrainer.cpp b/src/training/unicharset/lstmtrainer.cpp
index b608942e..183cc57c 100644
--- a/src/training/unicharset/lstmtrainer.cpp
+++ b/src/training/unicharset/lstmtrainer.cpp
@@ -38,6 +38,9 @@
 #endif
 #include "tprintf.h"

+#include <filesystem>
+namespace fs = std::filesystem;
+
 namespace tesseract {

 // Min actual error rate increase to constitute divergence.
@@ -1063,6 +1066,16 @@ std::vector<int> LSTMTrainer::MapRecoder(const UNICHARSET &old_chset,
   return code_map;
 }

+// Initializes the character set encode/decode mechanism directly from a
+// previously setup traineddata containing dawgs, UNICHARSET and
+// UnicharCompress. Note: Call before InitNetwork!
+void LSTMTrainer::InitCharSet(const std::string &traineddata_path) {
+  if (!std::filesystem::exists(traineddata_path.c_str()))
+    tprintf("Traineddata path %s not exists!\n", traineddata_path.c_str());
+  ASSERT_HOST(mgr_.Init(traineddata_path.c_str()));
+  InitCharSet();
+}
+
 // Private version of InitCharSet above finishes the job after initializing
 // the mgr_ data member.
 void LSTMTrainer::InitCharSet() {
diff --git a/src/training/unicharset/lstmtrainer.h b/src/training/unicharset/lstmtrainer.h
index cdebf989..bca8d43b 100644
--- a/src/training/unicharset/lstmtrainer.h
+++ b/src/training/unicharset/lstmtrainer.h
@@ -95,10 +95,8 @@ public:
   // Initializes the character set encode/decode mechanism directly from a
   // previously setup traineddata containing dawgs, UNICHARSET and
   // UnicharCompress. Note: Call before InitNetwork!
-  void InitCharSet(const std::string &traineddata_path) {
-    ASSERT_HOST(mgr_.Init(traineddata_path.c_str()));
-    InitCharSet();
-  }
+  void InitCharSet(const std::string &traineddata_path);
+
   void InitCharSet(const TessdataManager &mgr) {
     mgr_ = mgr;
     InitCharSet()
```

**Comment by stweil** [comment]
Fixed in commit 68017dbf2a584730adc6c562fd6afc722c957fcb.
